browser/extensions/api/ftx_api.cc

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/extensions/api/ftx_api.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "brave/browser/ftx/ftx_service_factory.h"
+#include "brave/browser/profiles/profile_util.h"
+#include "brave/common/extensions/api/ftx.h"
+#include "brave/components/ftx/browser/ftx_service.h"
+#include "brave/components/ftx/browser/regions.h"
+#include "brave/components/ftx/common/pref_names.h"
+#include "brave/components/ntp_widget_utils/browser/ntp_widget_utils_region.h"
+#include "chrome/browser/extensions/api/tabs/tabs_constants.h"
+#include "chrome/browser/extensions/extension_tab_util.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/prefs/pref_service.h"
+#include "extensions/browser/extension_util.h"
+#include "extensions/common/constants.h"
+
+namespace {
+
+FTXService* GetFTXService(content::BrowserContext* context) {
+  return FTXServiceFactory::GetInstance()->GetForProfile(
+      Profile::FromBrowserContext(context));
+}
+
+bool IsFTXAPIAvailable(content::BrowserContext* context) {
+  return brave::IsRegularProfile(context);
+}
+
+}  // namespace
+
+namespace extensions {
+namespace api {
+
+ExtensionFunction::ResponseAction FtxGetFuturesDataFunction::Run() {
+  if (!IsFTXAPIAvailable(browser_context())) {
+    return RespondNow(Error("Not available in Tor/incognito/guest profile"));
+  }
+
+  auto* service = GetFTXService(browser_context());
+  bool data_request = service->GetFuturesData(
+      base::BindOnce(&FtxGetFuturesDataFunction::OnFuturesData, this));
+
+  if (!data_request) {
+    return RespondNow(Error("Could not make request for futures data"));
+  }
+
+  return RespondLater();
+}
+
+void FtxGetFuturesDataFunction::OnFuturesData(const FTXFuturesData& data) {
+  base::ListValue result;
+
+  for (const TokenPriceData& currency : data) {
+    auto point = std::make_unique<base::Value>(base::Value::Type::DICTIONARY);
+    point->SetStringKey("symbol", currency.symbol);
+    point->SetDoubleKey("price", currency.price);
+    point->SetDoubleKey("percentChangeDay", currency.percentChangeDay);
+    point->SetDoubleKey("volumeDay", currency.volumeDay);
+    result.Append(std::move(point));
+  }
+
+  Respond(OneArgument(std::move(result)));
```

Comment:
this is just going to respond with an empty ListValue if there was an error. Seems like we should respond with an error here

Reply:
empty result handed on the frontend, but no error is shown even though it should. Follow-up here https://github.com/brave/brave-browser/issues/15852